src/framework/MOM_diag_buffers.F90

```diff
@@ -0,0 +1,421 @@
+  integer :: ie !< The end slot of the array i-direction
+  integer :: je !< The end slot of the array j-direction
+
+  integer, allocatable, dimension(:) :: ids  !< List of diagnostic ids whose slot corresponds to the row in the buffer
+  integer :: length = 0 !< The number of slots in the buffer
+
+  contains
+
+  procedure(a_grow), deferred :: grow !< Increase the size of the buffer
+  procedure, public :: check_capacity_by_id !< Check the size size of the buffer and increase if necessary
+  procedure, public :: set_horizontal_extents !< Define the horizontal extents of the arrays
+  procedure, public :: mark_available !< Mark that a slot in the buffer can be reused
+  procedure, public :: grow_ids !< Increase the size of the vector storing diagnostic ids
+  procedure, public :: find_buffer_slot !< Find the slot corresponding to a specific diagnostic id
+end type diag_buffer_base
+
+!> Dynamically growing buffer for 2D arrays.
+type, extends(diag_buffer_base), public :: diag_buffer_2d
+  real, public, allocatable, dimension(:,:,:) :: buffer !< The actual buffer to store data
+
+  contains
+
+  procedure, public :: grow => grow_2d !< Increase the size of the buffer
+  procedure, public :: store => store_2d !< Store a field in the buffer, increasing as necessary
+end type diag_buffer_2d
+
+!> Dynamically growing buffer for 3D arrays.
+type, extends(diag_buffer_base), public :: diag_buffer_3d ; private
+  integer :: ks !< The start slot in the k-dimension
+  integer :: ke !< The last slot in the k-dimension
+  real, public, allocatable, dimension(:,:,:,:) :: buffer !< The actual buffer to store data
+
+  contains
+
+  procedure, public :: set_vertical_extent !< Set the vertical extents of the buffer
+  procedure, public :: grow => grow_3d !< Increase the size of the buffer
+  procedure, public :: store => store_3d !< Store a field in the buffer, increasing as necessary
+end type diag_buffer_3d
+
+contains
+
+!> Signature for the grow methods on n-dimension diagnostic buffer types
+subroutine a_grow(this)
+  class(diag_buffer_base), intent(inout) :: this !< The diagnostic buffer
+end subroutine
+
+!> Mark a slot in the buffer as unused based on a diagnostic id. For example,
+!! the data in that slot has already been consumed and can thus be overwritten
+subroutine mark_available(this, id)
+  class(diag_buffer_base), intent(inout) :: this !< The diagnostic buffer
+  integer,                 intent(in)    :: id   !< The diagnostic id
+  integer :: slot
+
+  slot = this%find_buffer_slot(id)
+  this%ids(slot) = 0
+end subroutine mark_available
+
+!> Return the slot of the buffer corresponding to the diagnostic id
+pure function find_buffer_slot(this, id) result(slot)
+  class(diag_buffer_base), intent(in) :: this !< The diagnostic buffer
+  integer, intent(in) :: id !< The diagnostic id
+
+  integer, dimension(1) :: temp
+  integer :: slot !< The slot in the buffer corresponding to the diagnostic id
+
+  if(allocated(this%ids)) then
```

Comment:
Please add a space after the `if` here and elsewhere to read `if (...)` to follow the conventions in the MOM6 style guide at https://github.com/NOAA-GFDL/MOM6/wiki/Code-style-guide#white-space.

Reply:
Completed